### PR TITLE
#947 Adds Gen.file which randomly gives files from a given directory.

### PR DESCRIPTION
**Pull request by ashishkujoy**
Not sure of how to write test for this without being dependent on external file (test resource).

### CODE REVIEW

**Comment by LeoColman** [comment]
I suppose that there's no harm on creating a `temp` file for this? But yes, it will be dependant.

But perhaps you can use any of kotlintest files?
For example, list all the files from the root directory, and then generate files from that directory and verify that all files were obtainable and that only files from the directory were obtained

File: kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt [hunk 1]
```diff
@@ -153,3 +148,25 @@ inline fun <reified T : Enum<T>> Gen.Companion.enum(): Gen<T> = object : Gen<T>
 
 fun Gen.Companion.regex(regex: String) = RegexpGen(regex)
 fun Gen.Companion.regex(regex: Regex) = regex(regex.pattern)
+
+/**
+ * Returns a stream of values where each value is a randomly
+ * chosen File object from given directory. Directory does not exists, an empty sequence will be return.
```
Comment (LeoColman) [inline kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt]:
```suggestion
 * chosen File object from given directory. If the directory does not exist, an empty sequence will be returned instead.
```

File: kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt [hunk 2]
```diff
@@ -153,3 +148,25 @@ inline fun <reified T : Enum<T>> Gen.Companion.enum(): Gen<T> = object : Gen<T>
...
+ * If recursive is true it gives files from inner directories as well recursively.
+ */
+fun Gen.Companion.file(directoryName: String, recursive: Boolean = false): Gen<File> = object : Gen<File> {
```
Comment (LeoColman) [inline kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt:157]: I wonder if this should be `Gen.file` or `Gen.files`. @sksamuel ?

File: kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt [hunk 3]
```diff
@@ -153,3 +148,25 @@ inline fun <reified T : Enum<T>> Gen.Companion.enum(): Gen<T> = object : Gen<T>
...
+   override fun constants(): Iterable<File> = emptyList()
+   override fun random(seed: Long?): Sequence<File> {
+      val fileTreeWalk = File(directoryName).walk()
+      return if (!recursive) {
```
Comment (LeoColman) [inline kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt]: I believe that this inverted condition is bad for code readability. Perhaps it's better to make it a `if (recursive)`?

File: kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt [hunk 4]
```diff
@@ -153,3 +148,25 @@ inline fun <reified T : Enum<T>> Gen.Companion.enum(): Gen<T> = object : Gen<T>
...
+         randomiseFiles(fileTreeWalk.maxDepth(1), seed)
+      } else randomiseFiles(fileTreeWalk, seed)
+   }
+
+   private fun randomiseFiles(files: Sequence<File>, seed: Long?): Sequence<File> {
+      val allFiles = files.toList()
+      if(allFiles.isEmpty()) return emptySequence()
+      val r = if (seed == null) Random.Default else Random(seed)
+      return generateInfiniteSequence { allFiles[r.nextInt(0, allFiles.size)] }
```
Comment (LeoColman) [inline kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt]: I believe you can use `allFiled.random(random)` instead, can't you?

**Review by LeoColman — CHANGES_REQUESTED** [review]

**Comment by LeoColman** [comment]
@sksamuel Did we change back for `random` to receive a `long` instead of a `kotlin.random`?

File: kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt (hunk 2 above)
Comment (sksamuel) [reply kotlintest-assertions/src/jvmMain/kotlin/io/kotlintest/properties/jvmgens.kt:157]: I think either is fine.

**Pull request merged**